inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -2,33 +2,1223 @@
+            Xbyak::Ymm, Xbyak::Zmm>::type;
+    size_t vlen = cpu_isa_traits<isa>::vlen;
+
+    Xbyak::Address table_val(int index) { return ptr[reg_table + index * vlen]; }
+    Xbyak::Address table_seq_val(int index) { return ptr[reg_table + jcp_.axis_dim * vlen + index * sizeof(float)]; }
+
+    Xbyak::Reg64 reg_src = r8;
+    Xbyak::Reg64 reg_dst = r9;
+    Xbyak::Reg64 reg_dst_idx = r10;
+    Xbyak::Reg64 reg_prc = r11;
+    Xbyak::Reg64 reg_prc_idx = r12;
+    Xbyak::Reg64 reg_work_amount = r13;
+    Xbyak::Reg64 reg_table = r14;
+    Xbyak::Reg64 reg_params = abi_param1;
+    Xbyak::Reg64 reg_j = rax;
+    Xbyak::Reg64 reg_aux = rdx;
+    Xbyak::Reg64 reg_aux_idx = rbx;
+
+    Xbyak::Reg8 reg_tmp_8 = r15b;
+    Xbyak::Reg32 reg_tmp_32 = r15d;
+    Xbyak::Reg64 reg_tmp_64 = r15;
+
+    Vmm vmm_zero = Vmm(1); // vmm_zero represents Vmm(1) when isa is avx512_common, otherwise vmm_mask represents Vmm(1)
+
+    const Xbyak::Opmask k_mask = Xbyak::Opmask(1);
+
+    unsigned char cmp_flg;
+    unsigned char heap_cmp_flg;
+
+    Xbyak::Label l_table;
+
+    std::unique_ptr<jit_emu_vcvtneps2bf16> emu_vcvtneps2bf16;
+
+    inline void topk_loop() {
+        if (jcp_.algorithm == TopKAlgorithm::topk_bubble_sort) {
+            if (jcp_.layout != TopKLayoutType::topk_blocked) {
+                topk_bubble_vector();
+                topk_bubble_scalar();
+            } else {
+                if (!jcp_.topk_innermost) {
+                    topk_bubble_vector();
+                } else {
+                    if (jcp_.top_k == 1) {
+                        topk_bubble_horiz_blocked_innermost();
+                    } else {
+                        topk_bubble_scalar_blocked_innermost();
+                    }
+                }
+            }
+        } else if (jcp_.algorithm == TopKAlgorithm::topk_bitonic_sort) {
+            if (jcp_.layout != TopKLayoutType::topk_blocked) {
+                topk_bitonic_vector();
+                topk_bitonic_scalar();
+            } else {
+                if (!jcp_.topk_innermost) {
+                    topk_bitonic_vector();
+                } else {
+                    topk_bitonic_vector_blocked_innermost();
+                    topk_bitonic_scalar_blocked_innermost();
+                }
+            }
+        } else if (jcp_.algorithm == TopKAlgorithm::topk_heap_sort) {
+            topk_heap_scalar();
+        }
+    }
+
+    inline void topk_bitonic_vector() {
+        Xbyak::Label topk_main_loop_label;
+        Xbyak::Label topk_main_loop_end_label;
+        int step = vlen / sizeof(float);
+        L(topk_main_loop_label);
+        {
+            cmp(reg_work_amount, step);
+            jl(topk_main_loop_end_label, T_NEAR);
+
+            // src => prc
+            for (int i = 0; i < jcp_.axis_dim; i++) {
+                load_vector(vmm_tmp, ptr[reg_src + i * jcp_.sort_stride * jcp_.data_size], jcp_.data_type);
```

Comment:
As far as I know it is decided to use jit_load_store_emitters instead of load_vector and store_vector when implementing new nodes.
See mvn and roi_pooling nodes for more information.

Reply:
Fixed. jit_load_store_emitters do helps a lot especially when dealing with tail elements.
load_vector and store_vector are all replaced by jit_load_store_emitters.
load_scalar and store_scalar are partly replaced. Because in some cases(eg. bitonic sort in blocked layout on innermost dimension, heap sort in planar layout on innermost dimension), element by element sorting is needed.
Thanks Egor!